### PR TITLE
feat: graceful rollback

### PR DESCRIPTION
**Pull request by jmhrpr**
# Scrolls Graceful Rollback-Handling Prototype

### Problem
Scrolls currently handles rollbacks by maintaining a buffer which withholds new blocks until the buffer has reached a configurable size (`min_depth` number of blocks), after which the oldest block is removed from the buffer and processed ([read more here](https://txpipe.github.io/oura/advanced/rollback_buffer.html)). That is, blocks are held for some time before Scrolls processes them so that rollbacks to points within the buffer can be easily handled because the blocks have not been processed yet. This has a trade-off: if you set the min-depth high rollbacks can be handled and the data in Scrolls won't be corrupted _but_ the data in Scrolls will lag behind the chain by `min_depth` number of blocks, on the other hand if you set the min-depth so low that a rollback is received for a block that Scrolls has already been processed then Scrolls is unable to revert any of the changes which were made by that block (and any processed blocks which succeeded it which have also been rollbacked). This means Scrolls cannot process blocks as soon as it receives them (`min-depth` of zero) without data being corrupted every time a rollback is received - the changes which were made by blocks which were later rolled-back will persist, rendering data affected by the block inaccurate.

### Goal
Modify Scrolls such that it can have the most up-to-date chain data/no-latency, whilst also guaranteeing that the data in Scrolls will not be corrupted by rollbacks. Do this by implementing a way to essentially process blocks in reverse/undo the effects of blocks.

### TLDR Solution

Add a rollback-buffer to the enrich stage which stores the recently processed blocks (points) along with consumed UTxOs (including the encoded output) which were deleted from the enrich-db and the produced UTxOs which were inserted, so they can be re-inserted or removed accordingly if the block is roll-backed.

Also add a rollback-buffer to storage stage, which stores the recently processed blocks (points) along with all the storage-altering commands submitted by the reducers stage _and_ the return values from the storage for each of these commands. When a block needs to be undone we use the commands and their corresponding return values to craft the inverse storage-altering commands and execute them against the storage to undo them.

### Detailed Solution
First let's discuss how a block is processed in Scrolls, so we can see what effects need to be reversed in order to "undo" the effects of a block which has since been rollbacked. There are 3 stages to processing an incoming block and they currently work as follows:
1. Enrich stage: a database mapping UTxO IDs to their values/controlling-address is maintained, and for each block being processed the consumed UTxOs are removed from the database and produced UTxOs are inserted into the database. The block and information about the UTxOs used in the block are sent to the reducer stage.
2. Reducer stage: the block is passed to the reducers configured by the user, which send commands to the next stage in order to modify the storage. For example, the TxByHash reducer sends a command to the storage stage to set the key `<tx_hash>` to the value `<tx_cbor_bytes>` for each transaction in the block.
3. Storage stage: this stage receives the commands from the reducer stage as they are sent and executes them against the storage backend being used (Redis).

So at a high-level the solution which can reverse the effects of rollbacked blocks should look like:
1. Receive a `RollbackToBlock(P)` message from the node
2. Figure out which blocks we have processed since the point `P`
3. For every block we have processed since `P` reverse the effects it made by:
        a. Reversing the changes it made to the UTxO database at the enrich stage by re-inserting the consumed UTxOs which we removed when processing the block, and removing the produced UTxOs we inserted when processing the block.
        b. Reversing the changes all the reducers made to the Scrolls storage (Redis)

It is not sufficient to simply store each block we process and attempt to "process it in reverse" when we want to undo it because the actions we take when processing a block are often destructive/the block does not contain all the information which is required to reverse the action, for example:

- In the enrich stage UTxO database is a key/value database where the key is UTxO ID (TxIn) and the value is the UTxO amount and controlling-address (TxOut). When processing a block we take all the transaction inputs (TxIns) and delete the entries in the database with these keys, as these UTxOs have been consumed. The problem is if we wanted to process this block in reverse in order to undo this change the only information we get from the block is the _keys_ of the deleted entries, but the block does not contain information about the _value_ of the deleted entries so we don't have enough information to reinsert the key-value entry. That is, the block contains sufficient information to process in one direction but not enough information to know how to undo it.

- The reducer stage/storage-commands face a similar problem in that some commands are destructive in nature. For example consider the command `set key K1 to value A`: to undo this operation we need to know if `K1` was already set to some other value `B` before the command was executed in which case the command to reverse the action would be `set key K1 to value B`, but if `K1` did not point to any value before the command then the reverse action would be `delete key K1`.

Instead we only store exactly the information required to undo any changes which were made as a result of processing each block. Each stage has a rollback buffer which maps blocks to the information required to reverse the effects of that block within the context of the stage:

In the enrich stage this means the rollback buffer is a list of recently processed block IDs mapping to the UTxOs and their corresponding values which were removed from the enrich DB (consumed by transactions in the block) and the UTxOs which were inserted into the enrich DB (produced by transactions in the block). Then, when the enrich stage receives a message to rollback to a point, it retrieves entries from the buffer for the blocks which were processed after the rollback point and for each block re-inserts the UTxOs into the enrich DB which were removed and deletes the UTxOs which were inserted.

In the storage stage the rollback buffer functions identically but instead maps block IDs to the storage-modifying commands and their corresponding return values/extra information which is required to craft the inverse storage-modifying commands. When the storage stage needs to rollback to a point, it retrieves the entries from the buffer for the blocks which were processed after the rollback point and for every storage command which was executed as a result of processing the rollback points we take the storage command (`set key K1 to value A`) and the corresponding return value (`value B overwritten`) and crafts the reverse storage (`set key K1 to value B`) command and execute it against storage. This allows us to undo each individual storage command. The code which reverses an action given an action and return value is found [here](https://github.com/maestro-org/scrolls/blob/13e36c587048a62703ddec9af524cf51070ae8b7/src/storage/redis.rs#L59-L110).

### Implementation
The rollback buffer structure is defined over a generic type `<T>`. This rollback buffer is a vector with elements of type `(Point, <T>)`, where the `Point` is the block and `<T>` is some structure which contains all the information required to undo the effects which resulted from processing the block `Point` in the given stage. For example in the enrich stage `<T>` = `(Vec<ConsumedUtxos>, Vec<ProducedUtxos>)` and in the storage stage (Redis) `<T>` = `Vec<(StorageAction, StorageActionReturnValue)>`.

In the enrich stage, when processing a `MsgRollforward(P)` we note all the UTxOs and their encoded outputs that we deleted from the DB and note all the UTxOs we inserted into the DB, then at the end of processing the block in the stage we push `(P, (deleted, inserted))` onto the front of the rollback buffer then send the block down the pipeline as normal. When processing a `MsgRollforward(P, A)` in the storage stage we execute all the actions `A` against the storage in a Redis transaction, collect the return values for each action `R` and then push `(P, zip(A, R))` onto the front of the rollback buffer (instead of crafting the reverse actions and storing them in the buffer we just store `zip(A, R)` and use that to craft the reverse actions if we need to because the block has rolled back, because rollbacks are rare so it makes sense to only spend time computing these when we need to).

In order to be able to pair storage actions with their response values (which were previously not collected) we change the reducer logic from sending actions directly to storage stage to instead adding the actions to a list of actions which is then sent in its entirety to the storage stage at the end of processing a block. This has the added future benefit of potentially having some method to merge related storage actions to reduce the number of storage operations which are required (i.e `PNCounter(K, 1) + PNCounter(K, 1) = PNCounter(K, 2)`).

Initially I envisaged having one rollback buffer which maps points with the information to reverse the block at _every_ stage, and then this information would be passed to the stages which could act on this information, but because the stages are separate workers which pass data in one direction there was no obvious way to pass data backwards through the stages (for example return the storage action return values back up to a previous stage where the global rollback buffer could reside), and so that it why each stage maintains its own rollback-buffer containing stage-specific information.

### Todo
- [X] implementation
- [X] method to test, via simulated chainsync rollforwards and rollbackwards or otherwise (*crude emulate source implemented for now*)
```
[source]
type = "Emulate"
instructions = "FFFB1" // go forward 3 blocks then rollback 1 block (test blocks located in sources/emulate/block_data/)
```
- [ ] persist and use the rollback buffers in such a way that Scrolls can still benefit them if the chain rolls-back while Scrolls is not running/has crashed
- [ ] don't waste time on rollback buffer when syncing data
- [ ] configurable rollback buffer size
- [ ] GC for increment-able keys
- [ ] Warn/Error on SortedSetAdds which weren’t performed (no overwrites allowed)
- [ ] TBD...

### CODE REVIEW

**Comment by jmhrpr** [comment]
Moving to https://github.com/maestro-org/scrolls/pull/4
